storagenode/piecestore/endpoint.go

```diff
@@ -649,7 +649,9 @@ func (endpoint *Endpoint) Retain(ctx context.Context, retainReq *pb.RetainReques
 
 		hasMorePieces = (len(pieceIDs) == limit)
 		offset += len(pieceIDs)
-		offset -= numDeleted
+		if endpoint.config.RetainStatus == RetainEnabled {
+			offset -= numDeleted
```

Comment:
This would still would be problematic when the deletion fails, so we need to track the pieceid where we left off, somehow... rather than using a numberic offset.